src/graphql/node/control.rs

```diff
@@ -39,172 +38,222 @@ impl NodeControlMutation {
+    async fn apply_node(&self, ctx: &Context<'_>, id: ID) -> Result<ApplyNodeResponse> {
         let i = id.as_str().parse::<u32>().map_err(|_| "invalid ID")?;
-        let agents = ctx.data::<BoxedAgentManager>()?;
 
-        let (node, _invalid_agents) = {
+        let (node, invalid_agents) = {
             let store = crate::graphql::get_store(ctx).await?;
             let node_map = store.node_map();
             node_map
                 .get_by_id(i)?
-                .ok_or_else(|| async_graphql::Error::new(format!("Node with ID {i} not found",)))?
+                .ok_or_else(|| async_graphql::Error::new(format!("Node with ID {i} not found")))?
         };
 
-        if node.name_draft.is_none() && node.profile_draft.is_none() {
-            return Err("There is nothing to apply.".into());
+        if !invalid_agents.is_empty() {
+            return Err(async_graphql::Error::new(format!(
+                "Node {i} cannot be applied due to invalid agents: {invalid_agents:?}"
+            )));
         }
 
-        let config_setted_modules = send_set_config_requests(agents, &node).await;
-        let success_modules = if let Ok(config_setted_modules) = config_setted_modules {
-            update_node(ctx, i, node.clone(), &config_setted_modules).await?;
+        if !valid_node_apply(&node) {
+            return Err("Node is not valid for apply".into());
+        }
 
-            if let Some(customer_id) = should_broadcast_customer_change(&node) {
-                broadcast_customer_change(customer_id, ctx).await?;
-            }
-            config_setted_modules
-        } else {
-            return Err("Failed to apply node profile".into());
-        };
+        let giganto_draft = node
+            .giganto
+            .as_ref()
+            .and_then(|g| g.draft.as_ref().map(ToString::to_string));
 
-        Ok(ApplyResult {
-            id,
-            success_modules,
-        })
+        let apply_scope = node_apply_scope(&node);
+
+        if let Some(ref target_agents) = apply_scope.agents {
+            let agent_manager = ctx.data::<BoxedAgentManager>()?;
+            notify_agents(
+                agent_manager,
+                &target_agents.updates,
+                &target_agents.disables,
+            )
+            .await?;
+
+            info!(
+                "[{}] Node ID {i} - Node's agents are notified to apply.\n{:?}",
+                chrono::Utc::now(),
+                node.agents.iter().filter_map(|agent| {
+                    if target_agents.updates.contains(&agent.key.as_str()) {
+                        Some(format!(
+                            "\nAgent key: {}, Config: {:?}, Draft: {:?}",
+                            agent.key, agent.config, agent.draft
+                        ))
+                    } else {
+                        None
+                    }
+                })
+            );
+        }
+
+        if apply_scope.db {
```

Comment:
I wonder why notifying the agents precedes updating the db. I can see that this might cause an error, which could prevent the database from being updated.

Reply:
In a separate meeting, the following points were synced on this:

- Conclusion: We will update the database first and then notify the agents. We will not perform DB rollback actions within the `applyNode` API function due to the failures in agent notification.
- Background for the Conclusion
  - It was previously agreed that "An agent provides a function to indicate whether it is running properly without any failures during the (re)loading of an (updated) configuration."
  - In this context, "running properly" encompasses not only the agent being operational (or alive) but also whether it is functioning with the expected configuration values. Since verifying if the agents are running with the correct configurations requires a separate function that checks and compares the expected and actual configurations, a rollback decision within `applyNode` would not be reliable.
  - Moreover, the `applyNode` API will return a response based on the success or failure of the database update, independent of whether agent notifications succeed.

Reply:
You've changed this part according to my suggestion.